### PR TITLE
k0s status

### LINKED ISSUE #544: Make "k0s install" a bit more verbose
**Is your feature request related to a problem? Please describe.**
When running `k0s install` it would be useful to have some more information:

* The type of service installed
* The location of the startup script
* An output of the startup command

### PR DESCRIPTION
**Pull request by jasmingacic**
---
name: Pull Request
about: Create a Pull Request
title: 'k0s status'
labels: 'enhancememt'
assignees: ''

---

**Issue**
Fixes #544 #602 

**What this PR Includes**
A simple command that prints out following:
```
Process ID: 1076667
Role: server+worker
Init System: linux-systemd
```
of if --out parameter is used it will print json/yaml output
```
{
   "Pid": 1076667,
   "PPid": 856280,
   "Role": "server+worker",
   "SysInit": "linux-systemd",
   "StubFile": ""
}
```
```
pid: 1076667
ppid: 856280
role: server+worker
sysinit: linux-systemd
stubfile: ""
```

This is just a draft as I want to hear more input on what we would/should surface with this command

### CODE REVIEW

**Comment by jasmingacic** [comment]
I haven't tested this with worker nodes or any other OSs but linux. If anyone has k0s running as worker or on windows host please test it.

**Comment by ncopa** [comment]
What happens if there are multiple instances of k0s running on the same machine?

I wonder if it would be better to query the running instance via a unix socket or similar, rather than do assumptions based on what is found by the process list.

**Comment by jasmingacic** [comment]
@ncopa valid question.  Current setup wouldn't be able to detect all running instances. We could perhaps try looking if the binary is running as a service because we can't have two running services with the same name.

**Comment by ncopa** [comment]
@jasmingacic I think it would make more sense to connect to the running instance via an API. Then we we can also show state of supervised process (running/restarting) and do a health check of the components that supports that. Currently only etcd I think.

**Comment by jasmingacic** [comment]
@ncopa we briefly discussed the option you suggested [here](https://github.com/k0sproject/k0s/pull/637#issuecomment-762959720) and as @jnummelin suggested there might be some security concerns. Also currently it isn't not possible to run two k0s instances on the same node.
This command is essentially just convenience for some work that needs to be done on k0sctl.

File: cmd/status.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+				PPid: p.PPid()}
+
+			return status, nil
+		}
+	}
+
+	return &K0sStatus{}, nil
+}
+
+func hasChildern(pid int, processes []ps.Process) bool {
```
Comment (jnummelin) [inline cmd/status.go:161]: `hasChildern ` -> `hasChildren `

File: cmd/status.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+}
+func getRole(pid int) (role string, err error) {
+	if runtime.GOOS == "windows" {
+		return "worker", nil
+	}
+
+	var raw []byte
+	if raw, err = ioutil.ReadFile(fmt.Sprintf("/proc/%d/cmdline", pid)); err != nil {
+		return "", fmt.Errorf("K0s not running")
```
Comment (jnummelin) [inline cmd/status.go:132]: the most obvious reason for this to fails is not running the command as root, not k0s not running

File: cmd/status.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	output string
+
+	status    *K0sStatus
+	statusCmd = &cobra.Command{
+		Use:     "status",
+		Short:   "Helper command for get general information about k0s",
+		Example: `The command will return information about system init, PID, k0s role, kubeconfig and similar.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var err error
+
```
Comment (jnummelin) [inline cmd/status.go:44]: I think we should check if we're running this as root or not. we're looking for stuff like process list, `/proc/1234/cmdline` which will not be readable without root privileges

File: cmd/status.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	var raw []byte
+	if raw, err = ioutil.ReadFile(fmt.Sprintf("/proc/%d/cmdline", pid)); err != nil {
+		return "", fmt.Errorf("K0s not running")
+	}
+	cmdln := string(raw)
+	if strings.Contains(cmdln, "enable-worker") {
+		return "server+worker", nil
+	} else if strings.Contains(cmdln, "server") {
+		return "server", nil
+	}
```
Comment (jnummelin) [inline cmd/status.go:139]: Missing case to detect pure worker

File: cmd/status.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	return role, nil
```
Comment (jnummelin) [inline cmd/status.go:140]: I don't think it makes sense to return empty string as role as a "fallback"

**Review by jnummelin — CHANGES_REQUESTED** [review]

**Comment by jasmingacic** [comment]
```
# ./k0s status --out json
{
   "Version": "v0.9.0-41-g1e859ee",
   "Pid": 0,
   "PPid": 0,
   "Role": "",
   "SysInit": "",
   "StubFile": ""
}
```
```
# ./k0s status --out yaml
version: v0.9.0-41-g1e859ee
pid: 0
ppid: 0
role: ""
sysinit: ""
stubfile: ""
```
```
# ./k0s status 
Version: v0.9.0-41-g1e859ee
K0s not running
```

File: cmd/status.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	for _, p := range processes {
+		if p.PPid() == pid {
+			return true
+		}
+	}
+	return false
+}
+
+func getProcessOwner(pid int) (string, error) {
+	stdout, err := exec.Command("ps", "-o", "user=", "-p", strconv.Itoa(pid)).Output()
```
Comment (kke) [inline cmd/status.go:192]: I think you could just use the [process.Username](https://pkg.go.dev/github.com/shirou/gopsutil/process#Process.Username)

File: cmd/status.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	Pid      int
+	PPid     int
+	Role     string
+	SysInit  string
+	StubFile string
+	output   string
+}
+
+func (s K0sStatus) String() {
+	s.Version = build.Version
```
Comment (kke) [inline cmd/status.go:93]: This needs to be the version of the running daemon, not the version of current executable.

Comment (jasmingacic) [reply cmd/status.go:93]: @jnummelin do you think we can achieve this in any other way?

Comment (kke) [reply cmd/status.go:93]: 1) write stuff to a file under the run-dir
2) `/proc/$pid/exe version`

Comment (kke) [reply cmd/status.go:93]: method 1 would be nice, k0s could pretty much write a `K0sStatus` json to rundir `status.json`

**Comment by jasmingacic** [comment]
Once approved I will squash the commits

**Review by kke — APPROVED** [review]
Ok for now.

The slightly hackish ways to determine information about the running process should probably be replaced with some sort of `status.json` in the run-dir in the future which would be written out when k0s starts, deleted when it ends and maybe kept up to date for some parts during k0s lifetime.

File: cmd/status.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	default:
+		if s.Pid == 0 {
+			fmt.Println("K0s not running")
+			return
+		}
+
+		fmt.Println("Version:", s.Version)
+		fmt.Println("Process ID:", s.Pid)
+		fmt.Println("Parent Process ID:", s.PPid)
+		fmt.Println("Role:", s.Role)
```
Comment (kke) [inline cmd/status.go:114]: (could just default to yaml output and drop these, yaml is human readable)

**Review by jnummelin — APPROVED** [review]

**Pull request merged**